### DISCUSSION TITLE
How to install pyvista with vtk_osmesa?

### DISCUSSION THREAD

**Discussion by nknavasto** [pyvista/pyvista#5597] [Q&A] [answered]
I am building a wheel for my python package and would like to include `pyvista` as a dependency. However, this wheel is supposed to be running on a headless environment. So instead of `vtk` as its dependency, I would like to use `vtk_osmesa` instead. 

My current approach is to
- checkout pyvista
- change the settings in `pyproject.toml`
- compile a `pyvista_osmesa` package
- upload that package to my own python package index and
- specify `pyvista_osmesa` as a dependency.

My questions in this board are:

1. Is there a way to specify `pyvista `as a dependency and use `vtk_osmesa ` instead?
2. Is my current approach violating any license terms?

**Comment by MatthewFlamm** [comment]
I think trying to maintain a customized version of pyvista will be a lot of work and pain in the long run.  Does this recommendation work for you? https://discourse.vtk.org/t/status-update-vtk-python-wheels/11212


```txt
pip install pyvista  # use local cloned copy if you need it
pip uninstall vtk -y
pip install --extra-index-url https://wheels.vtk.org vtk-osmesa
```

Unfortunately, this requires installing vtk, uninstalling it, then installing vtk-osmesa.  This problem occurs, in my view, because Python does not allow having a default optional dependency.  ~Another factor is that Kitware has decided to rename the package `vtk-osmesa~

**Reply by nknavasto** [reply]
@MatthewFlamm Thank you for your response. 
This is also a way I had in mind. The issue is, that the customers who install my package then need to type multiple lines in a console. I am already scared, that they need to call one line (pip install <my-package>)
Would it be possible to make `vtk` as an optional dependency? In the `requirements.txt` file, it is already excluded! How about doing something like

```
pip install pyvista[osmesa]
```

Then the dependency would be `vtk_osmesa` instead of `vtk`. I think a consequence of this approach is, that `pip install pyvista` then would not install any `vtk` version. But I am not 100% sure here.`

**Reply by MatthewFlamm** [reply]
I understand what you are trying to do now, thanks for the added info.  This is a tricky one.

> Would it be possible to make vtk as an optional dependency? In the requirements.txt file, it is already excluded!

The problem in your suggestion is that for pyvista only users this won't work: `pip install pyvista`.  We would need `pip install pyvista[vtk]`.  As you are afraid of users having to execute multiple lines, it is too much to ask users to use an optional dependency for the default usage in my opinion.  It is a hard problem.

One thought that might be a lot of burden here is to have a separate `pyvista-osmesa` package just like vtk has separate packages.

**Reply by MatthewFlamm** [reply]
One more though on having a separate `pyvista-osmesa` package is that you'd have to specify a separate wheel location for vtk-osmesa since it is not available on pypi.  So users would still have to do something like `pip install pyvista-osmesa --extra-index-url https://wheels.vtk.org`.  It is still not ideal.

**Reply by nknavasto** [reply]
That is all correct what you are mentioning. Currently, I would add the requirement in my `setup.py` as follows:

```
setup(
...
install_requires=["pyvista_osmesa @ https: ..."]
)
```

This has the drawback, that an installation based on local python wheels (without internet access) is not possible any longer.

My favorite solution would be for you to provide a build with `pyvista-osmesa`, just like `vtk` is doing. Do you thing there is a chance for you to do that? I think the build pipeline would only have to be altered minimally (change the name of the package and the vtk dependency).

**Reply by banesullivan** [reply]
+1 to everything @MatthewFlamm has said here

Though I do want to add that I have some build configurations that make building and maintaining your own wheel for VTK quite easy and this is what I do these days. Check out https://github.com/banesullivan/vtk-cmake

**Reply by jules-l-jimmy** [reply]
Hi,
Will something like `pip install pyvista-osmesa --extra-index-url https://wheels.vtk.org` be integrated in the code as mentioned by @MatthewFlamm ?

Would appreciate to have a way to add a pyvista with vtk-osmesa to a requirements file

Thank you

**Reply by natsuwater** [reply, marked as answer]
If you can use vtk-9.4, it seems that just `pip install vtk` works:

https://discourse.vtk.org/t/status-update-runtime-opengl-render-window-selection-in-vtk/14583
